[clang][Sema] Fix for enums overflowing (#24667)

wheatman opened the pull request:
Enums which do not have a specified type can only grow to bigger types which contain more bits than the prior types.  This means that the largest signed integer type cannot grow to the largest unsigned integer types.

Fixes #24667

wheatman committed:
[clang][Sema] Fix for overflow in enumerators(#24667)

Enums which do not have a specified type can only grow to bigger types
which contain more bits than the prior types.  This means that the
largest signed integer type cannot grow to the largest unsigned integer types.

In the process also implements N3029 Improved Normal Enumerations and
patially implements N3030 Enhancements to Enumerations which brings C enums
more inline with C++ enums.

Fixes #24667

wheatman commented:
@cor3ntin I tried to ask for your review since it was just what github recommended, if you can't do it, could you please suggest somebody else.
Thanks.

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -19807,20 +19807,46 @@ EnumConstantDecl *Sema::CheckEnumConstant(EnumDecl *Enum,
+          Diag(IdLoc, diag::err_enumerator_wrapped)
+              << toString(EnumVal, 10) << EltTy;
+
+        } else if (T.isNull()) {
+          if (EltTy->isSignedIntegerType() &&
+              (getLangOpts().CPlusPlus ||
+               LangStandard::getLangStandardForKind(getLangOpts().LangStd)
+                   .isC23())) {
```
```suggestion
        } else if (T.isNull()) {
          if (EltTy->isSignedIntegerType() &&
              (getLangOpts().CPlusPlus ||  getLangOpts().C23)) {
```

wheatman replied:
Thanks you, I did see that in my looking

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -19807,20 +19807,46 @@ EnumConstantDecl *Sema::CheckEnumConstant(EnumDecl *Enum,
+            } else {
+              assert(false && "Enum Type is not basic integral type");
+            }
+          } else {
```
Can you use `getCorrespondingUnsignedType` here?

wheatman replied:
Thanks you, I did see that in my looking

cor3ntin reviewed (commented):
For the C changes, we should make sure we implement the relevant changes of 
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3029.htm and 
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3030.htm

wheatman commented:
> For the C changes, we should make sure we implement the relevant changes of https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3029.htm and https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3030.htm

In reference to these I am not sure, my reading of these indicates that c is basically taking on the c++ rules with more careful language, which I think I do by checking for either c++ or c23 and above, but would be happy to hear a better interpretation of the c rules

wheatman commented:
@AaronBallman I am a little confused about what is the correct behavior for one the the cases in the test file for c.
```cpp
enum PR4515 {PR4515a=1u,PR4515b=(PR4515a-2)/2};
int CheckPR4515[PR4515b==0?1:-1];
```
At present since everything is ints this works fine, but some things are not clear to me once the enum is allowed to be an unsigned type
https://godbolt.org/z/8bfr8s3n5

```cpp
enum PR4515 {PR4515a=1u,PR4515b=(PR4515a-2)/2};
 auto a = PR4515a;
 auto b = PR4515b;
 auto c = (1 - 2)/2;
 auto d = (1u - 2)/2;
 auto e = (PR4515a-2)/2;
```
values
```
a:
        .long   1                               # 0x1

b:
        .long   0                               # 0x0

c:
        .long   0                               # 0x0

d:
        .long   2147483647                      # 0x7fffffff

e:
        .long   0                               # 0x0
```


specifically around the value for `PR4515b=(PR4515a-2)/2` 
if `PR4515a` has type unsigned int, then the value of this expression is 2147483647 see `d`, but if it has type int then it has value of 0 see `c` which is the expected value for the test.  However, with the new change the enum type would be unsigned since it was specified as unsigned, which I believe is the new behavior.

AaronBallman commented:
> @AaronBallman I am a little confused about what is the correct behavior for one the the cases in the test file for c.
> 
> ```c++
> enum PR4515 {PR4515a=1u,PR4515b=(PR4515a-2)/2};
> int CheckPR4515[PR4515b==0?1:-1];
> ```
> 
> At present since everything is ints this works fine, but some things are not clear to me once the enum is allowed to be an unsigned type https://godbolt.org/z/8bfr8s3n5
> 
> ```c++
> enum PR4515 {PR4515a=1u,PR4515b=(PR4515a-2)/2};
>  auto a = PR4515a;
>  auto b = PR4515b;
>  auto c = (1 - 2)/2;
>  auto d = (1u - 2)/2;
>  auto e = (PR4515a-2)/2;
> ```
> 
> values
> 
> ```
> a:
>         .long   1                               # 0x1
> 
> b:
>         .long   0                               # 0x0
> 
> c:
>         .long   0                               # 0x0
> 
> d:
>         .long   2147483647                      # 0x7fffffff
> 
> e:
>         .long   0                               # 0x0
> ```
> 
> specifically around the value for `PR4515b=(PR4515a-2)/2` if `PR4515a` has type unsigned int, then the value of this expression is 2147483647 see `d`, but if it has type int then it has value of 0 see `c` which is the expected value for the test. However, with the new change the enum type would be unsigned since it was specified as unsigned, which I believe is the new behavior.

I think the enumeration constant type is `int` and not `unsigned int`, per 6.7.2.2p12 bullet 4: `int, if given explicitly with = and the value of the integer constant expression is representable by an int; or,`

```
enum PR4515 {
  PR4515a=1u,             // has type int per 6.7.2.2p12 bullet 4 because 1 is representable by an int
  PR4515b=(PR4515a-2)/2   // has type int per 6.7.2.2p12 bullet 4 because (1 - 2) / 2 is representable by an int
};
```
WDYT?

wheatman commented:
I believe I have correct implementations added for N3029 and N3030.  
I also tried to add tests.  One thing I was not sure about was how to add tests that have different expected behavior for different versions.  I just used  `#if __STDC_VERSION__  < 202311L` but please let me know if there is a better way.  I also updated the release notes to mention these additions.

shafik commented on clang/test/Sema/enum.c:
```diff
@@ -18,6 +19,40 @@ enum x                      // expected-warning {{enumeration values exceed rang
+        B = 42LL << 32,
+      C = -4, D = 12456 };
+
+enum g {  // too negative
```
What does this comment mean? There is no diagnostic here? This also applies to the `too pos` comment below.

wheatman replied:
I have removed those comments, those were copied showing the different behavior in different modes, but those comments should only be in the pre C23 version

shafik reviewed (commented):
I think a more appropriate title would be "Fix for overflow in enumerators"

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -135,6 +135,12 @@ C23 Feature Support
   ``__TYPE_FMTb__`` (e.g., ``__UINT_FAST64_FMTB__``) in C23 mode for use with
   macros typically exposed from ``<inttypes.h>``, such as ``PRIb8``.
   (`#81896: <https://github.com/llvm/llvm-project/issues/81896>`_).
+- Enumerations should allow values greater than INT_MAX and smaller than
+  INT_MIN, in order to provide a value-preserved set of integer constants. `N3029 Improved Normal Enumerations <https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3029.htm>`_
```
```suggestion
- Enumerations should allow values greater than ``INT_MAX`` and smaller than
  ``INT_MIN``, in order to provide a value-preserved set of integer constants. `N3029 Improved Normal Enumerations <https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3029.htm>`_
```

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -211,6 +217,8 @@ Bug Fixes in This Version
 - Clang now doesn't produce false-positive warning `-Wconstant-logical-operand`
   for logical operators in C23.
   Fixes (`#64356 <https://github.com/llvm/llvm-project/issues/64356>`_).
+- Fixes miscompilation when an enum has a specified value such that the auto
+  increment overflows a signed long. Fixes (`#24667 <https://github.com/llvm/llvm-project/issues/24667>`_)
```
```suggestion
- Fixes a miscompilation when an enum has a specified value such that the automatic
  increment overflows a ``signed long``. Fixes #GH24667.
```

AaronBallman commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -19797,6 +19797,19 @@ EnumConstantDecl *Sema::CheckEnumConstant(EnumDecl *Enum,
+          // the type of the integer constant expression, if given
+          // explicitly with = and if the value of the integer
+          // constant expression is not representable by int;
+          if (isRepresentableIntegerValue(Context, EnumVal, Context.IntTy)) {
+            Val = ImpCastExprToType(Val, Context.IntTy, CK_IntegralCast).get();
+          }
```
```suggestion
          if (isRepresentableIntegerValue(Context, EnumVal, Context.IntTy))
            Val = ImpCastExprToType(Val, Context.IntTy, CK_IntegralCast).get();
```
Per our odd coding standard.

AaronBallman reviewed (commented):
Thank you for working on this! One thing I would like to see are some tests added to `clang/test/C/C2x/` (`n3029.c` and `n3030.c`) that tests the examples and changes from the original papers/final standard to demonstrate the features are complete. There are other tests in that directory that show the boilerplate for C feature testing.

Aside from testing and some nits, I think this is pretty close to ready.

wheatman commented:
I made the changes, added the tests, and fixed some things that the tests found.

A few questions.
It seems like n3030 is only partially supported, specifically 
```
extern enum forward fwd_val0; /* Constraint violation: incomplete type */
extern enum forward* fwd_ptr0; /* Constraint violation: enums cannot be
                                  used like other incomplete types */
```
However, these seem unrelated to the rest of the changes and I am not really sure how to begin with them so I just labeled the support as partial for now.

Also, I did not touch anything on the msvc side due to this comment
```
      // For MSVC ABI compatibility, unfixed enums must use an underlying type
      // of 'int'. However, if this is an unfixed forward declaration, don't set
      // the underlying type unless the user enables -fms-compatibility. This
      // makes unfixed forward declared enums incomplete and is more conforming.
```
It seems like they don't support these features yet and I figured breaking matched behavior was not desirable.
How do I set up the tests to not run on the msvc side?

shafik commented:
ping, are you waiting on feedback on your last comment to move forward?
